From 989127f7b7d69b8ca33838d8b4db066d95e1d732 Mon Sep 17 00:00:00 2001
From: Jeff Moyer <jmoyer@redhat.com>
Date: Fri, 10 Oct 2014 21:03:33 -0400
Subject: [block] blk-mq: limit memory consumption if a crash dump is active

Message-id: <1412975015-5370-33-git-send-email-jmoyer@redhat.com>
Patchwork-id: 97456
O-Subject: [RHEL7 PATCH 32/34] blk-mq: limit memory consumption if a crash dump is active
Bugzilla: 1146660
RH-Acked-by: David Milburn <dmilburn@redhat.com>
RH-Acked-by: Mike Snitzer <snitzer@redhat.com>

This is a backport of the following commit.  This resolves bug
1146660.

  commit aedcd72f6c283dffefbb8b808ae67bdd2c6eb11a
  Author: Jens Axboe <axboe@fb.com>
  Date:   Wed Sep 17 08:27:03 2014 -0600

    blk-mq: limit memory consumption if a crash dump is active

    It's not uncommon for crash dump kernels to be limited to 128MB or
    something low in that area. This is normally not a problem for
    devices as we don't use that much memory, but for some shared SCSI
    setups with huge queue depths, it can potentially fill most of
    memory with tons of request allocations. blk-mq does scale back
    when it fails to allocate memory, but it scales back just enough
    so that blk-mq succeeds. This could still leave the system with
    not enough memory to make any real progress.

    Check if we are in a kdump environment and limit the hardware
    queues and tag depth.

    Signed-off-by: Jens Axboe <axboe@fb.com>

Signed-off-by: Jeff Moyer <jmoyer@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/block/blk-mq.c b/block/blk-mq.c
index be24273..39dc6e0 100644
--- a/block/blk-mq.c
+++ b/block/blk-mq.c
@@ -20,6 +20,7 @@
 #include <linux/cache.h>
 #include <linux/sched/sysctl.h>
 #include <linux/delay.h>
+#include <linux/crash_dump.h>
 
 #include <trace/events/block.h>
 
@@ -1741,6 +1742,16 @@ struct request_queue *blk_mq_init_queue(struct blk_mq_tag_set *set)
  if (!ctx)
   return ERR_PTR(-ENOMEM);
 
+ /*
+  * If a crashdump is active, then we are potentially in a very
+  * memory constrained environment. Limit us to 1 queue and
+  * 64 tags to prevent using too much memory.
+  */
+ if (is_kdump_kernel()) {
+  set->nr_hw_queues = 1;
+  set->queue_depth = min(64U, set->queue_depth);
+ }
+
  hctxs = kmalloc_node(set->nr_hw_queues * sizeof(*hctxs), GFP_KERNEL,
    set->numa_node);
 
-- 
1.7.1